core/model/model.go

```diff
@@ -40,39 +42,58 @@ type Model struct {
 	ModelType ModelType
 }
 
+// ValidateSeriesArgs holds the arguments for the ValidateSeries method
+type ValidateSeriesArgs struct {
+	ModelType ModelType
+	Name      string
+	Series    string
+	Format    charm.Format
+}
+
 var caasOS = set.NewStrings(os.Kubernetes.String())
 
 // ValidateSeries ensures the charm series is valid for the model type.
-func ValidateSeries(modelType ModelType, charmSeries string, charmFormat charm.Format) error {
-	if charmFormat >= charm.FormatV2 {
-		system, err := systems.ParseSystemFromSeries(charmSeries)
+func ValidateSeries(args ValidateSeriesArgs) error {
+	if args.Format >= charm.FormatV2 {
+		system, err := systems.ParseSystemFromSeries(args.Series)
 		if err != nil {
 			return errors.Trace(err)
 		}
-		if system.Resource != "" {
-			switch modelType {
-			case CAAS:
-				// CAAS models support using a resource as the system.
-				return nil
-			case IAAS:
-				return errors.NotValidf("IAAS models don't support systems referencing a resource")
-			}
-		}
-	} else {
-		os, err := series.GetOSFromSeries(charmSeries)
-		if err != nil {
-			return errors.Trace(err)
+
+		if system.Resource == "" {
+			return nil
 		}
-		switch modelType {
+
+		switch args.ModelType {
+		// CAAS models support using a resource as the system.
 		case CAAS:
-			if !caasOS.Contains(os.String()) {
-				return errors.NotValidf("series %q in a kubernetes model", charmSeries)
-			}
+			return nil
 		case IAAS:
-			if caasOS.Contains(os.String()) {
-				return errors.NotValidf("series %q in a non container model", charmSeries)
-			}
+			return errors.NotValidf("IAAS models don't support systems referencing a resource")
+		}
+
+		return nil
+	}
+
+	os, err := series.GetOSFromSeries(args.Series)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	switch args.ModelType {
+	case CAAS:
+		if !caasOS.Contains(os.String()) {
+			return errors.NewNotValid(nil, fmt.Sprintf(
+				`%q is not a Containers-as-a-Service (like kubernetes) charm`, args.Name,
```

Comment:
So "container charm" is sufficient here.
`NotValidf` appends "not valid" to the text, so I'd prefer us to use that along with the appropriate preamble, as before.

Reply:
Done & Done